controller/api/destination/external-workload/endpoints_controller.go

```diff
@@ -0,0 +1,513 @@
+				queueLatency: queueLatencyHist,
+			},
+		},
+		stop: stopCh,
+		log: logging.WithFields(logging.Fields{
+			"component": "external-endpoints-controller",
+		}),
+	}
+
+	_, err := k8sAPI.Svc().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ec.updateService,
+		DeleteFunc: ec.updateService,
+		UpdateFunc: func(_, newObj interface{}) {
+			ec.updateService(newObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = k8sAPI.ExtWorkload().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ew, ok := obj.(*ewv1alpha1.ExternalWorkload)
+			if !ok {
+				ec.log.Errorf("error processing ExternalWorkload event: expected *v1alpha1.ExternalWorkload, got %#v", obj)
+				return
+			}
+
+			if len(ew.Spec.WorkloadIPs) == 0 {
+				ec.log.Debugf("skipping ExternalWorkload event: %s/%s has no IP addresses", ew.Namespace, ew.Name)
+				return
+			}
+
+			if len(ew.Spec.Ports) == 0 {
+				ec.log.Debugf("skipping ExternalWorkload event: %s/%s has no ports", ew.Namespace, ew.Name)
+				return
+			}
+
+			services, err := ec.getSvcMembership(ew)
+			if err != nil {
+				ec.log.Errorf("failed to get service membership for %s/%s: %v", ew.Namespace, ew.Name, err)
+				return
+			}
+
+			for svc := range services {
+				ec.enqueueUpdate(svc)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			var ew *ewv1alpha1.ExternalWorkload
+			if ew, ok := obj.(*ewv1alpha1.ExternalWorkload); ok {
+				services, err := ec.getSvcMembership(ew)
+				if err != nil {
+					ec.log.Errorf("failed to get service membership for %s/%s: %v", ew.Namespace, ew.Name, err)
+					return
+				}
+
+				for svc := range services {
+					ec.enqueueUpdate(svc)
+				}
+				return
+			}
+
+			tomb, ok := obj.(cache.DeletedFinalStateUnknown)
+			if !ok {
+				ec.log.Errorf("error processing ExternalWorkload event: couldn't retrieve obj from DeletedFinalStateUnknown %#v", obj)
+				return
+			}
+
+			ew, ok = tomb.Obj.(*ewv1alpha1.ExternalWorkload)
+			if !ok {
+				ec.log.Errorf("error processing ExternalWorkload event: DeletedFinalStateUnknown contained object that is not a v1alpha1.ExternalWorkload %#v", obj)
+				return
+			}
+
+			services, err := ec.getSvcMembership(ew)
+			if err != nil {
+				ec.log.Errorf("failed to get service membership for %s/%s: %v", ew.Namespace, ew.Name, err)
+				return
+			}
+
+			for svc := range services {
+				ec.enqueueUpdate(svc)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			old, ok := oldObj.(*ewv1alpha1.ExternalWorkload)
+			if !ok {
+				ec.log.Errorf("error processing ExternalWorkload event: expected *v1alpha1.ExternalWorkload, got %#v", oldObj)
+			}
+
+			updated, ok := newObj.(*ewv1alpha1.ExternalWorkload)
+			if !ok {
+				ec.log.Errorf("error processing ExternalWorkload event: expected *v1alpha1.ExternalWorkload, got %#v", newObj)
+			}
+
+			// Ignore resync updates. If nothing has changed in the object, then
```

Comment:
Are you sure we want to ignore resyncs?  What if we miss an update, how will we recover?  E.g. suppose writing the EndpointSlice for a service fails and therefore an update to an ExternalWorkload is not reflected in the EndpointSlice.  How and when does the state get corrected?

Reply:
@adleong I think this is tricky for a couple of reasons. My initial recommendation when I chatted with @mateiidavid was to prune the updates (determine whether it is a material update as you said earlier). I now see this is all a bit more nuanced. The problem with resyncs is the following in my eyes. 

I am assuming we do not miss events coming from the watcher while the controller is running. What might happen is to error when we actually do the write to the k8s API and miss the update this way. A resync can fix that, but the problem is that on the resyncs you get an Update that has two identical old + new. So if you have failed to do an update where an ExternalWorkload transitions from selecting over a service to one that does not select this service anymore, your next resync will give you an Update that has old and new that do not select over the Service. 

In order to remediate the situation you would have to then go over the endpoints of the slices for all services to check whether there are any that should not be there (e.g. refering to this external workload when it should not).
Correct me if I am wrong but this seems rather excessive. This means that if you have 10 ExternalWorkloads in a namespace and 20 services on each resync you would need to do 200 of these full reconciliations. Does that seem correct? 

So two options in my mind: 

- process resyncs and do this full diffing on the resyncs
- have a routine when the controller starts, that fetches all ExternalWorklaods and services and reconciles the state  (to avoid the situation where you have missed updates while the controller was down). So a full state sync. After that if you do not want to deal with all resyncs but are worried about failing writes, couple the processing logic with a retriable queue with a delay (similar to the multicluster controller). This way you can requeue an update if it fails. 

Happy to hear more opinions on the matter.